File: Test/ModernTheme.cpp
#include "Application.h"
#include "Callback.h"
#include "ComboBox.h"
#include "ComboBoxItem.h"
#include "ConstraintLayout.h"
#include "Label.h"
#include "MainWindow.h"
#include "PopupMenu.h"

using namespace d14uikit;

#define DEMO_NAME L"ModernTheme"

int main(int argc, char* argv[])
{
    float dpi = 96.0f;
    if (argc >= 2 && strcmp(argv[1], "HighDPI") == 0)
    {
        dpi = 192.0f;
    }
    Application app(DEMO_NAME, dpi);
    app.setHeight(300);
    app.setMinSize(app.size());
    app.setResizable(true);

    //------------------------------------------- Initialize UI objects.

    MainWindow mwnd(DEMO_NAME);
    ConstraintLayout centerLayout;
    mwnd.setContent(&centerLayout);

    ConstraintLayout::GeoInfo geoInfo = {};

    Label themeLbl(L"Theme mode");
    themeLbl.setHeight(40);
    themeLbl.setHorzAlign(Label::HCenter);
    geoInfo = {};
    geoInfo.keepWidth = false;
    geoInfo.Left.ToLeft = 50;
    geoInfo.Right.ToLeft = 350;
    geoInfo.keepHeight = true;
    geoInfo.Top.ToTop = 100;
    centerLayout.addElement(&themeLbl, geoInfo);

    ComboBox themeSelector;
    themeSelector.setHeight(40);
    geoInfo = {};
    geoInfo.keepWidth = false;
    geoInfo.Left.ToRight = 350;
    geoInfo.Right.ToRight = 50;
    geoInfo.keepHeight = true;
    geoInfo.Top.ToTop = 100;
    centerLayout.addElement(&themeSelector, geoInfo);
    themeSelector.setRoundRadius(5);

    auto menu = themeSelector.dropDownMenu();

    ComboBoxItem items[3];
    items[0].setText(L"Light");
    items[1].setText(L"Dark");
    items[2].setText(L"Use system setting");

    std::list<MenuItem*> pItems;
    for (auto& item : items)
    {
        pItems.push_back(&item);
    }
    menu->appendItem(pItems);

    menu->setWidth(themeSelector.width());
    menu->setHeight(_countof(items) * 40);
    menu->setRoundExtension(5);

    themeSelector.setCurrSelected(2);
    
    //------------------------------------------- Set UI event callacks.

    themeSelector.D14_onSelectedChange(ComboBox, obj, text)
    {
        auto app = Application::app();
        if (text == L"Light" || text == L"Dark")
        {
            app->setThemeMode(text);
        }
        else if (text == L"Use system setting")
        {
            app->setUseSystemTheme(true);
        }
    };
    return app.run();
}

File: Test/PyBind/scripts/intermediate/ModernTheme.py
from sys import argv

from D14UIKit import *

DEMO_NAME = 'ModernTheme'

def main():
    dpi = 96.0
    if len(argv) >= 2 and argv[1] == 'HighDPI':
        dpi = 192.0

    app = Application(DEMO_NAME, dpi)
    app.height = 300
    app.minSize = app.size
    app.resizable = True

    #------------------------------------------- Initialize UI objects.

    mwnd = MainWindow(DEMO_NAME)
    centerLayout = ConstraintLayout()
    mwnd.content = centerLayout

    themeLbl = Label('Theme mode')
    themeLbl.height = 40
    themeLbl.horzAlign = Label.HCenter
    geoInfo = ConstraintLayout.GeoInfo()
    geoInfo.keepWidth = False
    geoInfo.Left.ToLeft = 50
    geoInfo.Right.ToLeft = 350
    geoInfo.keepHeight = True
    geoInfo.Top.ToTop = 100
    centerLayout.addElement(themeLbl, geoInfo)

    themeSelector = ComboBox()
    themeSelector.height = 40
    geoInfo = ConstraintLayout.GeoInfo()
    geoInfo.keepWidth = False
    geoInfo.Left.ToRight = 350
    geoInfo.Right.ToRight = 50
    geoInfo.keepHeight = True
    geoInfo.Top.ToTop = 100
    centerLayout.addElement(themeSelector, geoInfo)
    themeSelector.roundRadius = 5

    menu = themeSelector.dropDownMenu

    items = [ComboBoxItem() for i in range(3)]
    items[0].text = 'Light'
    items[1].text = 'Dark'
    items[2].text = 'Use system setting'

    menu.appendItem(items)

    menu.width = themeSelector.width
    menu.height = len(items) * 40
    menu.roundExtension = 5

    themeSelector.setCurrSelected(2)

    #------------------------------------------- Set UI event callbacks.

    def changeThemeMode(obj, text):
        app = Application.app
        if text == 'Light' or text == 'Dark':
            style = app.themeStyle
            style.name = text
            app.themeMode = style
        elif text == 'Use system setting':
            app.useSystemTheme = True

    themeSelector.f_onSelectedChange = changeThemeMode

    exit(app.run())

if __name__ == '__main__':
    main()
